cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -815,6 +815,233 @@ class AzureFileSystem::Impl {
+                    FileType::Directory};
+  }
+
+  static std::string_view BasenameView(std::string_view s) {
+    auto offset = s.find_last_of(internal::kSep);
+    auto tail = (offset == std::string_view::npos) ? s : s.substr(offset);
+    return internal::RemoveTrailingSlash(tail, /*preserve_root=*/false);
+  }
+
+  /// \brief List the blobs at the root of a container or some dir in a container.
+  ///
+  /// \pre container_client is the client for the container named like the first
+  /// segment of select.base_dir.
+  Status GetFileInfoWithSelectorFromContainer(
+      const Azure::Storage::Blobs::BlobContainerClient& container_client,
+      const Azure::Core::Context& context, Azure::Nullable<int32_t> page_size_hint,
+      const FileSelector& select, FileInfoVector* acc_results) {
+    ARROW_ASSIGN_OR_RAISE(auto base_location, AzureLocation::FromString(select.base_dir));
+
+    bool found = false;
+    Azure::Storage::Blobs::ListBlobsOptions options;
+    if (internal::GetAbstractPathDepth(base_location.path) == 0) {
+      // If the base_dir is the root of the container, then we want to list all blobs in
+      // the container and the Prefix should be empty and not even include the trailing
+      // slash because the container itself represents the `<container>/` directory.
+      options.Prefix = {};
+      found = true;  // Unless the container itself is not found later!
+    } else {
+      options.Prefix = internal::EnsureTrailingSlash(base_location.path);
+    }
+    options.PageSizeHint = page_size_hint;
+    options.Include = Azure::Storage::Blobs::Models::ListBlobsIncludeFlags::Metadata;
+
+    // When Prefix.Value() contains a trailing slash and we find a blob that
+    // matches it completely, it is an empty directory marker blob for the
+    // directory we're listing from, and we should skip it.
+    auto is_empty_dir_marker =
+        [&options](const Azure::Storage::Blobs::Models::BlobItem& blob) noexcept -> bool {
+      return options.Prefix.HasValue() && blob.Name == options.Prefix.Value();
+    };
+
+    auto recurse = [&](const std::string& blob_prefix) noexcept -> Status {
+      if (select.recursive && select.max_recursion > 0) {
+        FileSelector sub_select;
+        sub_select.base_dir = base_location.container;
+        sub_select.base_dir += internal::kSep;
+        sub_select.base_dir += internal::RemoveTrailingSlash(blob_prefix);
+        sub_select.allow_not_found = true;
+        sub_select.recursive = true;
+        sub_select.max_recursion = select.max_recursion - 1;
+        return GetFileInfoWithSelectorFromContainer(
+            container_client, context, page_size_hint, sub_select, acc_results);
+      }
+      return Status::OK();
+    };
+
+    // (*acc_results)[*last_dir_reported] is the last FileType::Directory in the results
+    // produced through this loop over the response pages.
+    std::optional<size_t> last_dir_reported{};
```

Comment:
Is this really needed?
The added tests are passed without this:

```diff
diff --git a/cpp/src/arrow/filesystem/azurefs.cc b/cpp/src/arrow/filesystem/azurefs.cc
index ffcab5755..75f3ea68c 100644
--- a/cpp/src/arrow/filesystem/azurefs.cc
+++ b/cpp/src/arrow/filesystem/azurefs.cc
@@ -913,15 +913,15 @@ class AzureFileSystem::Impl {
 
     // (*acc_results)[*last_dir_reported] is the last FileType::Directory in the results
     // produced through this loop over the response pages.
-    std::optional<size_t> last_dir_reported{};
-    auto matches_last_dir_reported = [&last_dir_reported,
-                                      acc_results](const FileInfo& info) noexcept {
-      if (!last_dir_reported.has_value() || info.type() != FileType::Directory) {
-        return false;
-      }
-      const auto& last_dir = (*acc_results)[*last_dir_reported];
-      return BasenameView(info.path()) == BasenameView(last_dir.path());
-    };
+    // std::optional<size_t> last_dir_reported{};
+    // auto matches_last_dir_reported = [&last_dir_reported,
+    //                                   acc_results](const FileInfo& info) noexcept {
+    //   if (!last_dir_reported.has_value() || info.type() != FileType::Directory) {
+    //     return false;
+    //   }
+    //   const auto& last_dir = (*acc_results)[*last_dir_reported];
+    //   return BasenameView(info.path()) == BasenameView(last_dir.path());
+    // };
 
     auto process_blob =
         [&](const Azure::Storage::Blobs::Models::BlobItem& blob) noexcept {
@@ -929,20 +929,21 @@ class AzureFileSystem::Impl {
             const auto& info = acc_results->emplace_back(
                 FileInfoFromBlob(base_location.container, blob));
             if (info.type() == FileType::Directory) {
-              last_dir_reported = acc_results->size() - 1;
+              // last_dir_reported = acc_results->size() - 1;
             }
           }
         };
     auto process_prefix = [&](const std::string& prefix) noexcept -> Status {
       const std::string path = base_location.container + internal::kSep + prefix;
       const auto& info = acc_results->emplace_back(DirectoryFileInfoFromPath(path));
-      if (ARROW_PREDICT_FALSE(matches_last_dir_reported(info))) {
-        acc_results->pop_back();
-      } else {
-        last_dir_reported = acc_results->size() - 1;
+      info.type();
+      // if (ARROW_PREDICT_FALSE(matches_last_dir_reported(info))) {
+        // acc_results->pop_back();
+      // } else {
+        // last_dir_reported = acc_results->size() - 1;
         return recurse(prefix);
-      }
-      return Status::OK();
+      // }
+      // return Status::OK();
     };
 
     try {
```

Reply:
This is needed because the Azure server might break the page after a "dir/", before a "dir/file".

I should add at least a comment explaining this and a unit test.

Reply:
It seems that a unit test for this case isn't added yet.